src/main/scala/goatrodeo/util/Helpers.scala

```diff
@@ -359,6 +373,28 @@ object Helpers {
     ???
   }
 
+  /** Compute the sha512 of an input stream
+    *
+    * @param in
+    *   the input stream
+    * @return
+    *   the bytes of the sha256 hash
+    */
+  def computeSHA512(in: InputStream): Array[Byte] = {
+    val md = MessageDigest.getInstance("SHA512")
+    val ba = new Array[Byte](4096)
+    while (true) {
+      val len = in.read(ba)
+      if (len <= 0) {
+        in.close()
+        return md.digest()
+      }
+
+      md.update(ba, 0, len)
+    }
+    ???
```

Comment:
This would throw a runtime exception.  Are we using this for some kind of boundary/break scenario?

Reply:
> Rust deals with this condition correctly

Whoa, interesting; how does Rust solve this?  Does Rust not have a bottom type?

Reply:
Rust does have a never type, `!`, which a function with an infinite loop will return

Reply:
It's a side effect (ha ha) of the Borrow Checker.

Because the borrow checker has to know what code can be executed and the ownership of references at each step, there's a lot of work that Rust does around stuff like `loop { }` or `while true { }`

Scala focuses on functional coding and the compiler has less reason to do full tracing.

This is a corner-case for Scala